### PR TITLE
CORE-19839 Add APIs for PublicKey encoding and decoding

### PR DESCRIPTION
**Pull request by bpaunescu**
Runtime OS PR is [here](https://github.com/corda/corda-runtime-os/pull/5798)

### CODE REVIEW

File: application/src/main/java/net/corda/v5/application/crypto/SigningService.java [hunk 1]
```diff
@@ -57,4 +57,44 @@ public interface SigningService {
     @Suspendable
     @NotNull
     Map<PublicKey, PublicKey> findMySigningKeys(@NotNull Set<PublicKey> keys);
+    
+    /**
+     * Deserializes a {@link PublicKey} from a {@code byte} array.
+     *
+     * @param encodedKey The public key represented asa a {@code byte} array.
```
Comment (simon-johnson-r3) [inline application/src/main/java/net/corda/v5/application/crypto/SigningService.java]: typo -> asa

File: application/src/main/java/net/corda/v5/application/crypto/SigningService.java [hunk 2]
```diff
@@ -57,4 +57,44 @@ public interface SigningService {
...
+     * @return An instance of {@link PublicKey} constructed from the encoded key.
+     */
+    @Suspendable
+    @NotNull
+    PublicKey decodePublicKey(@NotNull byte[] encodedKey);
+
+    /**
+     * Deserializes a {@link PublicKey} from a {@link String}.
+     *
+     * @param encodedKey The public key represented asa a hex encoded {@link String}.
```
Comment (simon-johnson-r3) [inline application/src/main/java/net/corda/v5/application/crypto/SigningService.java]: we refer to this as PEM encoded in the KeyEncodingService which I think is a bit more explanatory?

File: application/src/main/java/net/corda/v5/application/crypto/SigningService.java [hunk 3]
```diff
@@ -57,4 +57,44 @@ public interface SigningService {
...
+     *
+     * @param publicKey The {@link PublicKey} to be encoded.
+     * @return A {@code byte} array representation of the public key.
+     */
+    @Suspendable
+    @NotNull
+    byte[] encodeAsByteArray(@NotNull PublicKey publicKey);
+
+    /**
+     * Serializes a {@link PublicKey} into a hex encoded String.
```
Comment (simon-johnson-r3) [inline application/src/main/java/net/corda/v5/application/crypto/SigningService.java]: same here

**Review by simon-johnson-r3 — CHANGES_REQUESTED** [review]

File: application/src/main/java/net/corda/v5/application/crypto/SigningService.java [hunk 4]
```diff
@@ -57,4 +57,44 @@ public interface SigningService {
...
     @NotNull
     Map<PublicKey, PublicKey> findMySigningKeys(@NotNull Set<PublicKey> keys);
+    
+    /**
+     * Deserializes a {@link PublicKey} from a {@code byte} array.
+     *
+     * @param encodedKey The public key represented asa a {@code byte} array.
+     * @return An instance of {@link PublicKey} constructed from the encoded key.
+     */
+    @Suspendable
```
Comment (simon-johnson-r3) [inline application/src/main/java/net/corda/v5/application/crypto/SigningService.java]: In two minds whether these should be Suspendable, they will never actually suspend, but marking them as such opens up to future changes - however I see other functions in this module are not all marked as suspendable so maybe we should remove that?

File: application/src/main/java/net/corda/v5/application/crypto/SigningService.java (hunk 1 above)
Comment (bpaunescu) [reply application/src/main/java/net/corda/v5/application/crypto/SigningService.java]: Fixed.

File: application/src/main/java/net/corda/v5/application/crypto/SigningService.java (hunk 3 above)
Comment (bpaunescu) [reply application/src/main/java/net/corda/v5/application/crypto/SigningService.java]: Fixed.

File: application/src/main/java/net/corda/v5/application/crypto/SigningService.java (hunk 2 above)
Comment (bpaunescu) [reply application/src/main/java/net/corda/v5/application/crypto/SigningService.java]: Fixed.

File: application/src/main/java/net/corda/v5/application/crypto/SigningService.java (hunk 4 above)
Comment (bpaunescu) [reply application/src/main/java/net/corda/v5/application/crypto/SigningService.java]: Removed suspendable annotation.

**Review by simon-johnson-r3 — APPROVED** [review]

**Pull request merged**